opcua/client/client.py

```diff
@@ -38,27 +38,39 @@ def __init__(self, client, timeout):
         self.client = client
         self._dostop = False
         self._cond = Condition()
+        self._interruption_callbacks = set()
 
     def run(self):
         self.logger.debug("starting keepalive thread with period of %s milliseconds", self.timeout)
-        server_state = self.client.get_node(ua.FourByteNodeId(ua.ObjectIds.Server_ServerStatus_State))
-        while not self._dostop:
-            with self._cond:
-                self._cond.wait(self.timeout / 1000)
-            if self._dostop:
-                break
-            self.logger.debug("renewing channel")
-            self.client.open_secure_channel(renew=True)
-            val = server_state.get_value()
-            self.logger.debug("server state is: %s ", val)
-        self.logger.debug("keepalive thread has stopped")
+        try:
+            server_state = self.client.get_node(ua.FourByteNodeId(ua.ObjectIds.Server_ServerStatus_State))
+            while not self._dostop:
+                with self._cond:
+                    self._cond.wait(self.timeout / 1000)
+                if self._dostop:
+                    break
+                self.logger.debug("renewing channel")
+                self.client.open_secure_channel(renew=True)
+                val = server_state.get_value()
+                self.logger.debug("server state is: %s ", val)
+        except Exception as e:
+            self._connection_interrupted(e)
+        finally:
+            self.logger.debug("keepalive thread has stopped")
 
     def stop(self):
         self.logger.debug("stoping keepalive thread")
         self._dostop = True
         with self._cond:
             self._cond.notify_all()
 
+    def register_interruption_callback(self, callback):
+        self._interruption_callbacks.add(callback)
+
+    def _connection_interrupted(self, exception):
+        # copy the set for thread safety
+        for callback in self._interruption_callbacks.copy():
```

Comment:
If I understand you correctly, you mean something like:
    
    try:
        self.connection_interrupted = threading.Lock()
        self.connection_interrupted.acquire()
        while True:
            // send keepalive messages
    except Exception as e: // uh oh, connection lost
        self.connection_interrupted.release()

(or alternatively release the lock unconditionally)

and in the main thread you would do:

    client.connect()
    client.keepalive.connection_interrupted.aquire() # wait for connection loss
    # connection lost
    # do something about it

I see two problems: (1) the main thread cannot be used to interact with the opc ua client anymore (2) to fix this, one would have to call `connection_interrupted.aquire()` in a thread, in which case the *after connection loss*-code is called in a not-main thread too. Is there anything we gain by this?